WordPressLint/src/main/java/org/wordpress/android/lint/MissingNullAnnotationDetector.kt

```diff
@@ -0,0 +1,157 @@
+package org.wordpress.android.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.LintFix
+import com.android.tools.lint.detector.api.Scope.Companion.JAVA_FILE_SCOPE
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.isJava
+import com.intellij.psi.PsiPrimitiveType
+import org.jetbrains.uast.UAnnotated
+import org.jetbrains.uast.UAnnotationMethod
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UEnumConstant
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.UMethod
+import org.jetbrains.uast.UVariable
+
+class MissingNullAnnotationDetector: Detector(), SourceCodeScanner {
```

Comment:
1. Praise (❤️): Great work on creating this custom Android Lint rule!
2. Minor (🔍): This class can be automatically reformatted.
3. Suggestion (💡): I am not sure why some of the variables are outside the `MissingNullAnnotationDetector` class or its `companion object`, but I suggest moving them all inside for consistency purposes. Also, I suggest ordering and grouping them in such a way that it becomes easier to read through the class configuration, see example below:

<details>
    <summary>MissingNullAnnotationDetector.kt</summary>

```kotlin
package org.wordpress.android.lint

import com.android.tools.lint.client.api.UElementHandler
import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
import com.android.tools.lint.detector.api.Detector
import com.android.tools.lint.detector.api.Implementation
import com.android.tools.lint.detector.api.Issue
import com.android.tools.lint.detector.api.JavaContext
import com.android.tools.lint.detector.api.LintFix
import com.android.tools.lint.detector.api.Scope.Companion.JAVA_FILE_SCOPE
import com.android.tools.lint.detector.api.Severity
import com.android.tools.lint.detector.api.SourceCodeScanner
import com.android.tools.lint.detector.api.isJava
import com.intellij.psi.PsiPrimitiveType
import org.jetbrains.uast.UAnnotated
import org.jetbrains.uast.UAnnotationMethod
import org.jetbrains.uast.UElement
import org.jetbrains.uast.UEnumConstant
import org.jetbrains.uast.UField
import org.jetbrains.uast.UMethod
import org.jetbrains.uast.UVariable

class MissingNullAnnotationDetector : Detector(), SourceCodeScanner {
    override fun getApplicableUastTypes(): List<Class<out UElement>> = listOf(
        UField::class.java,
        UMethod::class.java,
    )

    override fun createUastHandler(context: JavaContext): UElementHandler? = with(context) {
        if (!isJava(uastFile?.sourcePsi)) {
            return null
        }
        object : UElementHandler() {
            override fun visitField(node: UField) {
                if (node.requiresNullAnnotation && !node.isNullAnnotated) {
                    report(node, MISSING_FIELD_ANNOTATION)
                }
            }

            override fun visitMethod(node: UMethod) {
                if (node.requiresNullAnnotation && !node.isNullAnnotated) {
                    report(node, MISSING_METHOD_RETURN_TYPE_ANNOTATION)
                }

                node.uastParameters.forEach { parameter ->
                    if (parameter.requiresNullAnnotation && !parameter.isNullAnnotated) {
                        report(parameter, MISSING_METHOD_PARAMETER_ANNOTATION)
                    }
                }
            }
        }
    }

    companion object {
        val MISSING_FIELD_ANNOTATION = Issue.create(
            id = "MissingNullAnnotationOnField",
            briefDescription = "Nullable/NonNull annotation missing on field",
            explanation = "Checks for missing `@NonNull/@Nullable` annotations for fields.",
        )
        val MISSING_METHOD_RETURN_TYPE_ANNOTATION = Issue.create(
            id = "MissingNullAnnotationOnMethodReturnType",
            briefDescription = "Nullable/NonNull annotation missing on method return type",
            explanation = "Checks for missing `@NonNull/@Nullable` annotations on method return types.",
        )
        val MISSING_METHOD_PARAMETER_ANNOTATION = Issue.create(
            id = "MissingNullAnnotationOnMethodParameter",
            briefDescription = "Nullable/NonNull annotation missing on method parameter",
            explanation = "Checks for missing `@NonNull/@Nullable` annotations on method parameters.",
        )

        /* ANNOTATED */

        private val UAnnotated.isNullAnnotated
            get() = this.uAnnotations.any { annotation ->
                acceptableNullAnnotations.any { nullAnnotation ->
                    annotation.qualifiedName == nullAnnotation
                }
            }

        @Suppress("ForbiddenComment")
        // TODO: Create a LintFix to normalize our usage to Android's annotations. This can be
        //  useful for running Nullability analysis. See:
        //  https://developer.android.com/studio/write/annotations#nullability-analysis
        private val acceptableNullAnnotations = listOf(
            "androidx.annotation.NonNull",
            "androidx.annotation.Nullable",
        )

        /* VARIABLE */

        private val UVariable.requiresNullAnnotation
            get() =
                !this.isPrimitive
                        && !this.isEnum
                        && !this.isConstant
                        && !this.isInitializedFinalField
                        && !this.isInjected

        private val UVariable.isPrimitive get() = this.type is PsiPrimitiveType
        private val UVariable.isEnum get() = this is UEnumConstant
        private val UVariable.isConstant get() = this.isStatic && this.isFinal
        private val UVariable.isInitializedFinalField
            get() = this.isFinal
                    && this.uastInitializer != null
        private val UVariable.isInjected
            get() = this.annotations.any { annotation ->
                annotation.qualifiedName?.endsWith("Inject") ?: false
            }

        /* METHOD */

        private val UMethod.requiresNullAnnotation
            get() =
                this !is UAnnotationMethod
                        && !this.isPrimitive
                        && !this.isConstructor
        private val UMethod.isPrimitive get() = this.returnType is PsiPrimitiveType

        /* CREATE */
       
        private fun Issue.Companion.create(
            id: String,
            briefDescription: String,
            explanation: String,
        ) = create(
            id = id,
            briefDescription = briefDescription,
            explanation = explanation,
            category = CORRECTNESS,
            priority = 5,
            severity = Severity.INFORMATIONAL,
            implementation = Implementation(
                MissingNullAnnotationDetector::class.java,
                JAVA_FILE_SCOPE,
            )
        )

        /* REPORT */

        private fun JavaContext.report(node: UElement, issue: Issue) = report(
            issue,
            node,
            getLocation(node),
            "Missing null annotation",
            node.fixes,
        )

        private val UElement.fixes
            get() = this.asSourceString().let { sourceString ->
                val nullableReplacement = "@Nullable $sourceString"
                val nonNullReplacement = "@NonNull $sourceString"

                LintFix.create().group()
                    .add(
                        LintFix.create()
                            .name("Annotate as @Nullable")
                            .replace()
                            .text(sourceString)
                            .shortenNames()
                            .reformat(true)
                            .with(nullableReplacement)
                            .build()
                    )
                    .add(
                        LintFix.create()
                            .name("Annotate as @NonNull")
                            .replace()
                            .text(sourceString)
                            .shortenNames()
                            .reformat(true)
                            .with(nonNullReplacement)
                            .build()
                    )
                    .build()
            }
    }
}
```

</details>

Reply:
>  I am not sure why some of the variables are outside the MissingNullAnnotationDetector class or its companion object, but I suggest moving them all inside for consistency purposes.

For one thing, the usages _within_ the companion object require that they are defined outside of it (e.g. the issue values created from the convenience `create` function). Also, [it is idiomatic to include extension functions at the top level](https://kotlinlang.org/docs/extensions.html#scope-of-extensions).

I considered extracting these to another file, but I'd favored not changing their their access modifiers - [so file-private seems most appropriate, imo](https://kotlinlang.org/docs/extensions.html#note-on-visibility).

> Also, I suggest ordering and grouping them in such a way that it becomes easier to read through the class configuration, see example below:

While I have already ordered and grouped them, I like the comments labelling the groups of extensions in lieu of extracting to separate files. That said, I'd prefer having the comment labels be based on the class being extended, for consistency.

Reply:
2: resolved in: https://github.com/wordpress-mobile/WordPress-Lint-Android/pull/11/commits/a1d6a7becf055c8f63022e99be360c29c72e427f

Reply:
Awesome, thanks for this update @mkevins and the comments labelling! 💯